### DISCUSSION TITLE
Real valued version of FFT

### DISCUSSION THREAD

**Discussion by lampuiho** [mathnet/mathnet-numerics#811] [Ideas]
Current available functions require complex input. Can you implement one that takes integers/float/double vector or array as input and output data in 2 input buffers, please?

See http://www.ti.com/lit/an/spra291/spra291.pdf for 2N-Point Real Sequence.

Also, for speed improvement, please allow the creation of a fourier transform object which pre-compute values of cosines and sines. (should be another ticket...)

**Comment by cdrnet** [comment]
We did indeed have such extensions in Iridium.

I do plan to work on FFT in a couple weeks, mostly on a native provider (fftw or MKL), but that would also be a chance to review real-valued use cases.

**Comment by mdabros** [comment]
Sounds great with some native FFT support. Will this also include 2D FFT?

**Comment by cdrnet** [comment]
Yes, that would be the idea.

**Comment by ThatGeoGuy** [comment]
If this is done, then this would also remedy my concerns laid out in issue #405, and the earlier issue created in #157. If done, will this functionality only be available in the MKL variant of Math.NET?

**Comment by cdrnet** [comment]
@lampuiho with `data in 2 input buffers` you refer to one array for the real and one for the imaginary part, just like the `void Forward(Double[] real, Double[] imaginary, FourierOptions options)` that was added in v3.14, but where the input content of the imaginary one would be ignored on forward transforms?

Meanwhile, in v3.14 I added a real version with a packed format (skipping information which is redundant due to the symmetries), see [Fourier.ForwardReal](http://numerics.mathdotnet.com/api/MathNet.Numerics.IntegralTransforms/Fourier.htm#ForwardReal).

**Comment by lampuiho** [comment]
Yes, exactly. But because what I needed was to compute FFT/IFFT for large amount of dataset (in a loop for different inputs) with very few data points, I ended up implementing my own real valued spit radix functions and unrolled function generator to avoid all those parallel calls which have too much overhead. Using external C++ library was also too slow for me because of the way C# handles memory.
